api/v1alpha3/reference_types.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import "k8s.io/apimachinery/pkg/runtime/schema"
+
+// ObjectReference contains enough information to let you inspect or modify the referred object.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type LocalObjectReference struct {
+	// API version of the referent.
+	// +optional
+	APIVersion string `json:"apiVersion,omitempty"`
```

Comment:
Do we want to use APIVersion here or just use APIGroup, similar to TypedLocalObjectReference? I think we'd likely want to avoid encoding the APIVersion to avoid the issues that exist here around migration/conversion/upgrade.

Reply:
@vincepri do you have a reference to a discussion where TypedLocalObjectReference is not meant for use outside of core types or is in the process of being deprecated?

Reply:
I don't. I vaguely remember something about it, but can't find anything related. 

From https://pkg.go.dev/k8s.io/api/core/v1?tab=doc#TypedLocalObjectReference:

```go
type TypedLocalObjectReference struct {
	// APIGroup is the group for the resource being referenced.
	// If APIGroup is not specified, the specified Kind must be in the core API group.
	// For any other third-party types, APIGroup is required.
	// +optional
	APIGroup *string `json:"apiGroup" protobuf:"bytes,1,opt,name=apiGroup"`
	// Kind is the type of resource being referenced
	Kind string `json:"kind" protobuf:"bytes,2,opt,name=kind"`
	// Name is the name of resource being referenced
	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
}
```

APIGroup is an optional string pointer, which isn't really what we're looking for, and having it empty it'd be implied to be part of the Kubernetes core, which seems odd for our uses as well.

I can see that the version might not be ideal, but I'm not sure if I prefer having it explicit and update it as we do today or not 🤔

Reply:
The updated proposed convention is in https://github.com/kubernetes/community/pull/4705